arch/nds/audio.c

```diff
@@ -0,0 +1,526 @@
+/* MegaZeux
+ *
+ * Copyright (C) 2016 Adrian Siekierka <asiekierka@gmail.com>
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License as
+ * published by the Free Software Foundation; either version 2 of
+ * the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
+ */
+
+#include "../../src/event.h"
+#include "../../src/platform.h"
+#include "../../src/util.h"
+#include "../../src/io/fsafeopen.h"
+#include "../../src/audio/audio.h"
+
+#include <nds.h>
+#include <maxmod9.h>
+#include "platform.h"
+#include "../../src/audio/audio.h"
+#include "../../src/audio/sfx.h"
+
+#ifdef CONFIG_AUDIO
+
+// Utility logic
+
+static u8 nds_vol_table[] = {
+	0, 8, 16, 26, 36, 48, 61, 75, 91, 108, 127
+};
+
+#define NDS_VOLUME(vol) nds_vol_table[CLAMP((vol), 0, 10)]
+
+// FIFO calls
+
+static inline void nds_sound_volume(int volume) {
+  fifoSendValue32(FIFO_MZX, CMD_MZX_SOUND_VOLUME | (NDS_VOLUME(volume) << 24));
+}
+
+static inline void nds_pcs_sound(int freq, int volume) {
+  fifoSendValue32(FIFO_MZX, CMD_MZX_PCS_TONE | (((freq) & 0xFFFF) << 8) | (NDS_VOLUME(volume) << 24));
+}
+
+// PC speaker implementation
+
+#define PCS_DURATION_PER_VBL 9
+
+static int pcs_playing;
+static int pcs_frequency;
+static int pcs_duration;
+
+static inline void nds_pcs_tick(int duration)
+{
+  int last_playing = pcs_playing;
+  int last_frequency = pcs_frequency;
+  int ticks = 0;
+  int ticked;
+
+  while (ticks < duration)
+  {
+    if (pcs_playing)
```

Comment:
while(
if(

There are plenty more of these but I'm not going to comment them all...